napari/components/viewer_model.py

```diff
@@ -602,9 +602,22 @@ def _on_remove_layer(self, event):
         layer = event.value
 
         # Disconnect all events from layer
-        layer.events.disconnect()
-        for em in layer.events.emitters.values():
-            em.disconnect()
+        layer.events.select.disconnect(self._update_active_layer)
+        layer.events.deselect.disconnect(self._update_active_layer)
+        layer.events.interactive.disconnect(self._update_interactive)
+        layer.events.cursor.disconnect(self._update_cursor)
+        layer.events.cursor_size.disconnect(self._update_cursor_size)
+        layer.events.data.disconnect(self._on_layers_change)
+        layer.events.scale.disconnect(self._on_layers_change)
+        layer.events.translate.disconnect(self._on_layers_change)
+        layer.events.rotate.disconnect(self._on_layers_change)
+        layer.events.shear.disconnect(self._on_layers_change)
+        layer.events.affine.disconnect(self._on_layers_change)
+        layer.events.name.disconnect(self.layers._update_name)
+
+        # layer.events.disconnect()
+        # for em in layer.events.emitters.values():
+        #     em.disconnect()
```

Comment:
Right, so if I understand this correctly, before, we were removing *all* connections, whether they were to this viewer or not. Now we only remove the ones to this viewer. Right?

This bit seems pretty hairy to me right now...

Reply:
Yes that's correct. What we do now is clearly not ok. Before we removed no connections, which caused lots of crashes, now we remove all connections which causes less bad things, and after this PR we'll remove exactly the right number of bad things :-)